transport/graphsync/requestqueue/requestqueue.go

```diff
@@ -0,0 +1,86 @@
+package requestqueue
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-graphsync"
+	"github.com/ipfs/go-peertaskqueue"
+	"github.com/ipfs/go-peertaskqueue/peertask"
+	ipld "github.com/ipld/go-ipld-prime"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+
+	datatransfer "github.com/filecoin-project/go-data-transfer"
+)
+
+const (
+	thawSpeed = time.Millisecond * 100
+)
+
+type RunDeferredRequestFn func(ctx context.Context, chid datatransfer.ChannelID, dataSender peer.ID, root ipld.Link, stor ipld.Node, doNotSendCids []cid.Cid, exts []graphsync.ExtensionData)
+type RequestQueue struct {
+	runDeferredRequest   RunDeferredRequestFn
+	maxInProcessRequests uint64
+
+	queue      *peertaskqueue.PeerTaskQueue
+	workSignal chan struct{}
+	ticker     *time.Ticker
+}
+
+func NewRequestQueue(runDeferredRequest RunDeferredRequestFn, maxInProcessRequests uint64) *RequestQueue {
+	return &RequestQueue{
+		maxInProcessRequests: maxInProcessRequests,
+		runDeferredRequest:   runDeferredRequest,
+		queue:                peertaskqueue.New(),
+		workSignal:           make(chan struct{}, 1),
+		ticker:               time.NewTicker(thawSpeed),
+	}
+}
+
+func (orq *RequestQueue) Start(ctx context.Context) {
+	for i := uint64(0); i < orq.maxInProcessRequests; i++ {
+		go orq.processRequestWorker(ctx)
+	}
+}
+
+func (orq *RequestQueue) processRequestWorker(ctx context.Context) {
```

Comment:
In this model there are many threads that all get "woken up" each time a task is added to the queue, and then all try to pop tasks from the queue, but most of the time all but one will get an empty task list.

I wonder if it would be possible to change the model so that there are initially zero go routines running. When a task is added to the queue a new go routine is started to process it. Each time a new go routine is needed it's started, up to some maximum number of go routines. As tasks complete, if the go routine is no longer needed it stops, all the way down to zero go routines.

This would allow us to use the minimum number of required go routines, and we would no longer need the `Start()` method.

I'm not sure if it's possible, but it would be nice if it is possible.